### PR TITLE
[staking] make contract staking indexer nullable

### PR DESCRIPTION
**Pull request by envestcc**
# Description
To improve flexibility and testability, make the contract staking indexer param in `staking.NewProtocol()` nullable.

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by bitcoinfinancier — APPROVED** [review]

File: action/protocol/staking/protocol.go [hunk 1]
```diff
@@ -475,7 +475,7 @@ func (p *Protocol) ActiveCandidates(ctx context.Context, sr protocol.StateReader
 	list := c.AllCandidates()
 	cand := make(CandidateList, 0, len(list))
 	for i := range list {
-		if ok && featureCtx.AddContractStakingVotes {
+		if p.contractStakingIndexer != nil && ok && featureCtx.AddContractStakingVotes {
```
Comment (Liuhaai) [inline action/protocol/staking/protocol.go:478]: Ok could be put at first

**Review by Liuhaai — APPROVED** [review]

File: action/protocol/staking/protocol.go [hunk 2]
```diff
@@ -475,7 +475,7 @@ func (p *Protocol) ActiveCandidates(ctx context.Context, sr protocol.StateReader
 	list := c.AllCandidates()
 	cand := make(CandidateList, 0, len(list))
 	for i := range list {
-		if ok && featureCtx.AddContractStakingVotes {
+		if ok && p.contractStakingIndexer != nil && featureCtx.AddContractStakingVotes {
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go:478]: 1. need to use `MustGetFeatureCtx` to make sure we can access the `AddContractStakingVotes` flag here
2. move if before for loop
```
if p.contractStakingIndexer != nil && protocol.MustGetFeatureCtx(ctx).AddContractStakingVotes {
   for i := range list
}
```

Comment (envestcc) [reply action/protocol/staking/protocol.go:478]: During service startup , `ActiveCandidates` is called at `nodeinfo/manager.go:104`, and the provided context does not contain the required feature. Using `MustGetFeatureCtx`  will lead to a panic.

Comment (dustinxie) [reply action/protocol/staking/protocol.go:478]: is it possible to add a feature context there?

Comment (dustinxie) [reply action/protocol/staking/protocol.go:478]: BTW, file a small separate PR for this

File: action/protocol/staking/staking_statereader.go [hunk 3]
```diff
@@ -75,9 +81,9 @@ func (c *compositeStakingStateReader) readStateBuckets(ctx context.Context, req
 	if err != nil {
 		return nil, 0, err
 	}
-
 	// merge native and LSD buckets
 	buckets.Buckets = append(buckets.Buckets, lsdIoTeXBuckets.Buckets...)
+
```
Comment (dustinxie) [inline action/protocol/staking/staking_statereader.go:86]: remove

File: chainservice/builder.go [hunk 4]
```diff
@@ -434,9 +435,15 @@ func (builder *Builder) buildNodeInfoManager() error {
...
+	chain := builder.cs.chain
 	dm := nodeinfo.NewInfoManager(&builder.cfg.NodeInfo, cs.p2pAgent, cs.chain, builder.cfg.Chain.ProducerPrivateKey(), func() []string {
-		candidates, err := stk.ActiveCandidates(context.Background(), cs.factory, 0)
+		ctx := protocol.WithFeatureCtx(
+			protocol.WithBlockCtx(
+				genesis.WithGenesisContext(context.Background(), chain.Genesis()),
+				protocol.BlockCtx{BlockHeight: chain.TipHeight()},
+			),
+		)
+		candidates, err := stk.ActiveCandidates(ctx, cs.factory, 0)
```
Comment (dustinxie) [inline chainservice/builder.go:446]: > only pass in stk.ActiveCandidates as a func param to NewInfoManager, later when it is called inside func (dm *InfoManager) updateBroadcastList(), you can construct the feature ctx using ch.TipHeight()

pls fix it that way, more clean and better readability

File: action/protocol/staking/protocol.go [hunk 5]
```diff
@@ -471,11 +471,10 @@ func (p *Protocol) ActiveCandidates(ctx context.Context, sr protocol.StateReader
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to get ActiveCandidates")
 	}
-	featureCtx, ok := protocol.GetFeatureCtx(ctx)
 	list := c.AllCandidates()
 	cand := make(CandidateList, 0, len(list))
 	for i := range list {
-		if ok && featureCtx.AddContractStakingVotes {
+		if p.contractStakingIndexer != nil && protocol.MustGetFeatureCtx(ctx).AddContractStakingVotes {
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go:477]:
```
featureCtx := protocol.MustGetFeatureCtx(ctx)
for range {
   if xxx && featureCtx.AddContractStakingVotes
}
```

**Review by dustinxie — APPROVED** [review]

**Pull request merged**